File: blackjack_cpp/src/tree_utils.cpp
#include "tree_utils.h"
#include <iostream>
#include <iomanip>
#include <sstream>
#include <variant>
#include <functional>

using namespace std;

namespace blackjack {

// Get all nodes organized by level
map<int, vector<AbstractBJTreeNode*>> getNodesByLevels(AbstractBJTreeNode* root) {
    map<int, vector<AbstractBJTreeNode*>> nodes_by_level;
    
    if (!root) return nodes_by_level;
    
    NodeIterator iter(root);
    while (iter.hasNext()) {
        NodeLevel node_level = iter.next();
        nodes_by_level[node_level.level].push_back(node_level.node);
    }
    
    return nodes_by_level;
}

// Count stages in a list of nodes
map<string, int> countStages(const vector<AbstractBJTreeNode*>& nodes) {
    map<string, int> stage_counter;
    
    for (AbstractBJTreeNode* node : nodes) {
        BJStage stage;
        
        // Check if it's a SimulationResultNode
        SimulationResultNode* sim_node = dynamic_cast<SimulationResultNode*>(node);
        if (sim_node) {
            stage = BJStage::ROUND_OVER;
        } else {
            stage = node->bj_round_.getStage();
        }
        
        string stage_name;
        
        if (stage == BJStage::PLAYER_CARD) {
            stage_name = "PLAYER_CARD";
            size_t num_children = node->children_.size();
            if (num_children == 1) {
                stage_name += "_SINGLE";
            } else if (num_children < 10) {
                stage_name += "_PARTIAL";
            } else {
                stage_name += "_FULL";
            }
        } else {
            // Convert BJStage enum to string
            switch (stage) {
                case BJStage::NOT_STARTED: stage_name = "NOT_STARTED"; break;
                case BJStage::PLAYER_CARD: stage_name = "PLAYER_CARD"; break;
                case BJStage::DEALER_CARD: stage_name = "DEALER_CARD"; break;
                case BJStage::PLAYER_ACTION: stage_name = "PLAYER_ACTION"; break;
                case BJStage::PLAYER_OFFERED_INSURANCE: stage_name = "PLAYER_OFFERED_INSURANCE"; break;
                case BJStage::PLAYER_OFFERED_EARLY_SURRENDER: stage_name = "PLAYER_OFFERED_EARLY_SURRENDER"; break;
                case BJStage::DEALER_CHECK_BJ: stage_name = "DEALER_CHECK_BJ"; break;
                case BJStage::ROUND_OVER: stage_name = "ROUND_OVER"; break;
                default: stage_name = "UNKNOWN"; break;
            }
        }
        
        stage_counter[stage_name]++;
    }
    
    return stage_counter;
}

// Log tree structure with stage counts per level
void logTreeStructure(AbstractBJTreeNode* root) {
    if (!root) {
        cout << "Tree is empty" << endl;
        return;
    }
    
    auto nodes_by_level = getNodesByLevels(root);
    
    cout << "\n=== Tree Structure by Level ===" << endl;
    
    for (const auto& [level, nodes] : nodes_by_level) {
        auto stage_counts = countStages(nodes);
        
        cout << "Level " << setw(2) << level << ": " 
             << setw(5) << nodes.size() << " nodes  ";
        
        // Print stage counts
        bool first = true;
        for (const auto& [stage_name, count] : stage_counts) {
            if (!first) cout << ", ";
            cout << stage_name << ":" << count;
            first = false;
        }
        cout << endl;
    }
    
    cout << endl;
}

// Format card rank value as string (T for 10, A for 11)
string formatCardValue(int value) {
    if (value == 10) return "T";
    if (value == 11) return "A";
    return std::to_string(value);
}

// Format a TransitionEvent as string
string formatEvent(const TransitionEvent& event) {
    if (holds_alternative<int>(event)) {
        return formatCardValue(get<int>(event));
    } else if (holds_alternative<PlayerAction>(event)) {
        return to_string(get<PlayerAction>(event));
    } else if (holds_alternative<DealerAction>(event)) {
        return to_string(get<DealerAction>(event));
    }
    return "UNKNOWN";
}

// Helper to get node type name
static string getNodeTypeName(AbstractBJTreeNode* node) {
    // Check specific types
    if (dynamic_cast<SimulationResultNode*>(node)) return "SimResult";
    if (dynamic_cast<FloorCeilValueNode*>(node)) return "FloorCeilValue";
    if (dynamic_cast<DecisionNode*>(node)) return "Decision";
    if (dynamic_cast<DealerCheckBJNode*>(node)) return "DealerCheckBJ";
    if (dynamic_cast<SplitNode*>(node)) return "Split";
    if (dynamic_cast<HitNode*>(node)) return "Hit";
    if (dynamic_cast<DoubleNode*>(node)) return "Double";
    if (dynamic_cast<AbstractFloorCeilNode*>(node)) return "FloorCeil";
    if (dynamic_cast<ValueNode*>(node)) return "FloorCeil";
    return "Unknown";
}

// Helper to get stage string
static string getStageString(AbstractBJTreeNode* node) {
    SimulationResultNode* sim_node = dynamic_cast<SimulationResultNode*>(node);
    if (sim_node) return "Terminal";
    
    FloorCeilValueNode* fc_val_node = dynamic_cast<FloorCeilValueNode*>(node);
    if (fc_val_node) return "FloorCeilPlaceholder";
    
    BJStage stage = node->bj_round_.getStage();
    switch (stage) {
        case BJStage::NOT_STARTED: return "NOT_STARTED";
        case BJStage::PLAYER_CARD: return "PLAYER_CARD";
        case BJStage::DEALER_CARD: return "DEALER_CARD";
        case BJStage::PLAYER_ACTION: return "PLAYER_ACTION";
        case BJStage::PLAYER_OFFERED_INSURANCE: return "PLAYER_OFFERED_INSURANCE";
        case BJStage::PLAYER_OFFERED_EARLY_SURRENDER: return "PLAYER_OFFERED_EARLY_SURRENDER";
        case BJStage::DEALER_CHECK_BJ: return "DEALER_CHECK_BJ";
        case BJStage::ROUND_OVER: return "ROUND_OVER";
        default: return "UNKNOWN";
    }
}

// Helper to format numeric value
static string formatValue(double val, bool has_value) {
    if (!has_value) return "N/A";
    ostringstream oss;
    oss << showpos << fixed << setprecision(3) << val;
    return oss.str();
}

// Get best action index for DecisionNode
static int getBestActionIdx(AbstractBJTreeNode* node) {
    DecisionNode* decision = dynamic_cast<DecisionNode*>(node);
    if (!decision) return -1;
    if (!decision->has_built_children_) return -1;
    
    for (size_t i = 0; i < decision->children_prob_.size(); ++i) {
        if (decision->children_prob_[i] == 1.0) {
            return static_cast<int>(i);
        }
    }
    return -1;
}

// Format a tree node line for display
string formatNodeLine(
    AbstractBJTreeNode* node,
    const TransitionEvent* event,
    bool is_best,
    double prob,
    int level,
    int indent_size
) {
    ostringstream oss;
    
    // Indent
    oss << string(indent_size * level, ' ');
    
    // Best marker
    oss << (is_best ? "* " : "  ");
    
    // Event
    if (event) {
        oss << "[" << formatEvent(*event) << "] ";
    }
    
    // Node type and stage
    string node_type = getNodeTypeName(node);
    string stage_str = getStageString(node);
    oss << node_type;  // << "(" << stage_str << ")";
    
    // Probability
    if (prob > 0.0) {
        oss << " p=" << fixed << setprecision(4) << prob;
    }
    
    // Values
    AbstractFloorCeilNode* fc_node = dynamic_cast<AbstractFloorCeilNode*>(node);
    ValueNode* sim_node = dynamic_cast<ValueNode*>(node);
    
    double value = 0.0;
    bool has_value = false;
    double floor_val = 0.0;
    double ceil_val = 0.0;
    bool has_floor_ceil = false;
    
    if (sim_node) {
        value = sim_node->getValue();
        has_value = true;
    } else if (fc_node) {
        try {
            value = fc_node->getValue();
            has_value = true;
            floor_val = fc_node->getFloorValue();
            ceil_val = fc_node->getCeilValue();
            has_floor_ceil = true;
        } catch (...) {
            has_value = false;
        }
    }
    
    oss << " EV=" << formatValue(value, has_value);
    if (has_floor_ceil) {
        oss << " [" << formatValue(floor_val, true) << ", " << formatValue(ceil_val, true) << "]";
    }
    
    // Decision info
    DecisionNode* decision = dynamic_cast<DecisionNode*>(node);
    if (decision) {
        if (decision->hasDecided()) {
            auto choice = decision->getDecisionChoice();
            if (choice.has_value()) {
                oss << " -> " << to_string(choice.value());
            }
        } else {
            oss << " (possible: ";
            bool first = true;
            for (PlayerAction action : decision->meaningful_actions_) {
                if (!first) oss << ", ";
                oss << to_string(action);
                first = false;
            }
            oss << ")";
        }
    }
    
    return oss.str();
}

// Format tree output as hierarchical string
string formatTreeOutput(AbstractBJTreeNode* root, int n_levels, int indent_size) {
    vector<string> lines;
    
    // Inner traversal function
    function<void(AbstractBJTreeNode*, const TransitionEvent*, bool, double, int)> traverse;
    traverse = [&](AbstractBJTreeNode* node, const TransitionEvent* event, bool is_best, double prob, int level) {
        lines.push_back(formatNodeLine(node, event, is_best, prob, level, indent_size));
        
        if (level >= n_levels) return;
        
        // Check for terminal nodes
        if (dynamic_cast<SimulationResultNode*>(node)) return;
        if (dynamic_cast<FloorCeilValueNode*>(node)) return;
        
        if (!node->has_built_children_ || node->children_.empty()) return;
        
        int best_idx = getBestActionIdx(node);
        
        for (size_t i = 0; i < node->children_.size(); ++i) {
            AbstractBJTreeNode* child = node->children_[i].get();
            const TransitionEvent* child_event = &node->children_events_[i];
            double child_prob = (i < node->children_prob_.size()) ? node->children_prob_[i] : 0.0;
            bool child_is_best = (static_cast<int>(i) == best_idx);
            traverse(child, child_event, child_is_best, child_prob, level + 1);
        }
    };
    
    traverse(root, nullptr, false, 0.0, 0);
    
    ostringstream result;
    for (size_t i = 0; i < lines.size(); ++i) {
        if (i > 0) result << "\n";
        result << lines[i];
    }
    return result.str();
}

// Print formatted tree output
void printTree(AbstractBJTreeNode* root, int n_levels, int indent_size) {
    cout << formatTreeOutput(root, n_levels, indent_size) << endl;
}

} // namespace blackjack

File: blackjack/tree_utils.py
from collections import defaultdict, deque
from collections import Counter
from typing import Generator
from blackjack.abstract_node import AbstractBJTreeNode, ValueNode
from blackjack.blackjack_round import BJStage
from blackjack.abstract_node import FloorCeilValueNode
from blackjack.abstract_node import ValueNode


def get_nodes_by_levels(root_node, nodes_by_level_dict=None):
    if nodes_by_level_dict is None:
        nodes_by_level_dict = defaultdict(list)
    queue = deque()
    queue.append((root_node, 0))
    while queue:
        node, node_level = queue.popleft()
        # handle case of a node with 2 parents, up to level 4
        if node_level < 4 and node in nodes_by_level_dict[node_level]:
            continue
        nodes_by_level_dict[node_level].append(node)
        for child in node.children:
            queue.append((child, node_level + 1))
    return nodes_by_level_dict



def iterate_nodes_by_levels(root_node) -> Generator[tuple[int, AbstractBJTreeNode], None, None]:
    nodes_by_level_dict = defaultdict(list)
    queue = deque()
    queue.append((0, root_node))
    while queue:
        node_level, node = queue.popleft()
        # handle case of a node with 2 parents, up to level 4
        if node_level < 4 and node in nodes_by_level_dict[node_level]:
            continue  
        nodes_by_level_dict[node_level].append(node)
        yield node_level, node
        for child in node.children:
            queue.append((node_level + 1, child))


def log_tree_structure(root_node):
    nodes_dict = get_nodes_by_levels(root_node)
    for lvl, nodes in nodes_dict.items():
        stage_count = count_stages(nodes)
        print(f"Level {lvl}: {len(nodes)} nodes", stage_count)


def count_stages(node_list):
    stage_counter = Counter()
    for node in node_list:
        if isinstance(node, ValueNode):
            stage = BJStage.ROUND_OVER
        else:
            stage = node.bj_round.get_stage()
        
        if stage == BJStage.PLAYER_CARD:
            stage_name = "PLAYER_CARD"
            if len(node.children) == 1:
                stage_name += "_SINGLE"
            elif len(node.children) < 10:
                stage_name += "_PARTIAL"
            else:
                stage_name += "_FULL"
        else:
            stage_name = stage.name
        stage_counter[stage_name] += 1
    return stage_counter


def format_tree_output(root_node, n_levels, indent_size=2):
    """
    Format a nice hierarchical output of the game tree.
    
    Parameters:
    - root_node: The root node of the tree
    - n_levels: Number of levels to display (0 = just root, 1 = root + children, etc.)
    - indent_size: Number of spaces per indentation level
    
    Returns:
    - Formatted string representation of the tree
    """
    from blackjack.floor_ceil_node import DecisionNode
    
    lines = []
    
    def get_node_type_name(node):
        """Get a short name for the node type."""
        class_name = node.__class__.__name__
        return class_name.replace("Node", "")
    
    def format_value(val):
        """Format a numeric value."""
        if val is None:
            return "N/A"
        return f"{val:+.3f}"
    
    def get_best_action_idx(node):
        """Get the index of the best action for a decision node."""
        if not isinstance(node, DecisionNode):
            return None
        if not node.has_built_children:
            return None
        # Best action is the one with probability 1
        for i, p in enumerate(node.children_prob):
            if p == 1:
                return i
        return None
    
    def format_event(event):
        """Format an event (action or card) for display."""
        if event is None:
            return ""
        if hasattr(event, 'name'):  # Enum
            return event.name
        if isinstance(event, int):  # Card rank value
            if event == 10:
                return "T"
            elif event == 11:
                return "A"
            else:
                return str(event)
        return str(event)
    
    def format_node_line(node, event=None, is_best=False, level=0):
        """Format a single node line."""
        indent = " " * (indent_size * level)
        marker = "* " if is_best else "  "
        
        # Get values
        try:
            value = node.get_value() if hasattr(node, 'get_value') else None
        except:
            value = None
        
        try:
            floor_val = node.get_floor_value() if hasattr(node, 'get_floor_value') else None
        except:
            floor_val = None
        
        try:
            ceil_val = node.get_ceil_value() if hasattr(node, 'get_ceil_value') else None
        except:
            ceil_val = None
        
        # Build the line
        node_type = get_node_type_name(node)
        event_str = f"[{format_event(event)}] " if event is not None else ""
        
        # Get stage info for non-value nodes
        if isinstance(node, ValueNode):
            stage_str = "Terminal"
        elif isinstance(node, FloorCeilValueNode):
            stage_str = "<FloorCeilPlaceholder>"
        else:
            stage = node.bj_round.get_stage()
            stage_str = stage.name if hasattr(stage, 'name') else str(stage)
        
        # Format values
        value_str = f"EV={format_value(value)}"
        if floor_val is not None and ceil_val is not None:
            value_str += f" [{format_value(floor_val)}, {format_value(ceil_val)}]"
        
        # Add decision info for DecisionNode
        extra_info = ""
        if isinstance(node, DecisionNode):
            if node.has_decided():
                extra_info = f" -> {node.decision_choice.name}"
            else:
                actions = [a.name for a in node.possible_actions]
                extra_info = f" (possible: {', '.join(actions)})"
        
        return f"{indent}{marker}{event_str}{node_type}({stage_str}) {value_str}{extra_info}"
    
    def traverse(node, event=None, is_best=False, level=0):
        """Recursively traverse and format the tree."""
        lines.append(format_node_line(node, event, is_best, level))
        
        if level >= n_levels:
            return
        
        if isinstance(node, ValueNode):
            return
        
        if not node.has_built_children or not node.children:
            return
        
        best_idx = get_best_action_idx(node)
        
        for i, (child, child_event) in enumerate(zip(node.children, node.children_events)):
            child_is_best = (i == best_idx)
            traverse(child, child_event, child_is_best, level + 1)
    
    traverse(root_node)
    return "\n".join(lines)


def print_tree(root_node, n_levels=2, indent_size=2):
    """Print the formatted tree output."""
    print(format_tree_output(root_node, n_levels, indent_size))
